### PR TITLE
Evenly distribute output grad  into all matching inputs for min/max/median

### PR DESCRIPTION
**Pull request by zasdfgbnm**
[bc-breaking note]. Previously, in case there were multiple max/min/median elements with the same value, gradient propagated only to the first element with this value, now gradient is evenly distributed between all the elements. This results in a minimum subnorm gradient. 
cc: @ngimel @mruberry

### CODE REVIEW

File: tools/autograd/derivatives.yaml [hunk 1]
```diff
@@ -670,7 +670,7 @@
   self: index_select_backward(grad, dim, indices, self.sizes(), keepdim)
 
 - name: max(Tensor self) -> Tensor
-  self: select_first_equal_backward(grad, self, result)
+  self: evenly_dispatch_backward(grad, self, result)
```
Comment (ngimel) [inline tools/autograd/derivatives.yaml:673]: bikeshedding: `evenly_distribute_backward`?

File: tools/autograd/templates/Functions.cpp [hunk 2]
```diff
@@ -726,20 +726,10 @@ Tensor _fused_dropout_backward(Tensor grad, Tensor mask, double p1m) {
...
-    grad_input.copy_(grad);
-  }
-  else {
-    grad_input.index_put_(at::chunk(first_value_idx, grad_input.dim()), grad);
-  }
-
-  return grad_input;
+Tensor evenly_dispatch_backward(Tensor grad, const Tensor & input, const Tensor & value) {
+  auto mask = (input == value);
+  auto count = mask.sum(input.scalar_type());
```
Comment (ngimel) [inline tools/autograd/templates/Functions.cpp:731]: For tensors with >2^24 elements that all happen to be max/min/median, this is going to be inaccurate, so maybe better leave in int64 or double, depending on which is faster?

File: tools/autograd/templates/Functions.cpp [hunk 3]
```diff
@@ -726,20 +726,10 @@ Tensor _fused_dropout_backward(Tensor grad, Tensor mask, double p1m) {
...
+  return at::zeros_like(input).masked_fill_(mask, grad / count);
```
Comment (albanD) [inline tools/autograd/templates/Functions.cpp:732]: I wonder if it would be faster to do `mask.to(input.scalar_type()) * (grad / count)` here?
Also it might be worth it to special case when `count=1` where we would be able to do something more efficient than `masked_fill_()`.

Comment (ngimel) [reply tools/autograd/templates/Functions.cpp:732]: Good point, not even `mask.to` because TensorIterator would take care of it, and on cuda type promotion will be implicit. `grad/count` has to be converted to input.scalar_type for this to work though.

Comment (zasdfgbnm) [reply tools/autograd/templates/Functions.cpp:732]: Whether to choose `*` vs `masked_fill` depend on CPU vs CUDA:

```python
import torch
for device in ['cpu', 'cuda']:
    t = torch.torch.randint(100, (1024, 1024, 64), device=device, dtype=torch.float)
    s = t.sum()
    m = t.max()
    mask = (t == m)
    go = t.new_tensor(1.)
    torch.cuda.synchronize()
    print(device)
    %timeit mask * (go / s); torch.cuda.synchronize() if device == 'cuda' else None
    %timeit torch.zeros_like(t).masked_fill_(mask, go / s); torch.cuda.synchronize() if device == 'cuda' else None
```
```
cpu
44.5 ms ± 199 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
25.2 ms ± 130 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
cuda
848 µs ± 16.6 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
1.08 ms ± 19.3 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```
so there should be some logic
```
if (cuda) {
  *
} else {
  masked_fill
}
```
and I don't think it should have a separate case for `s == 1`, because the only difference is `grad / count` vs `grad`, and `grad` is a scalar tensor. `scalar_tensor.item()` is not faster than `scalar_tensor / another_scalar_tensor`.

File: tools/autograd/templates/Functions.cpp (hunk 2 above)
Comment (zasdfgbnm) [reply tools/autograd/templates/Functions.cpp:731]: It is a scalar, so I don't think it makes any difference on int64 vs fp64. Let's just use the default (int64).

File: tools/autograd/derivatives.yaml (hunk 1 above)
Comment (zasdfgbnm) [reply tools/autograd/derivatives.yaml]: fixed

**Comment by zasdfgbnm** [comment]
Should be ready now. See my replies to reviews.

**Review by ngimel — APPROVED** [review]

**Comment by michaelklachko** [comment]
@zasdfgbnm @albanD is this also the case for kthvalue op?

**Comment by albanD** [comment]
`kthvalue` is different as it returns indices (just like `max(dim=)`) and so to be consistent, these ops only returns gradients for the index that was chosen during the forward.
So no I don't think this applies to the current kthvalue function.
If we had a version of kthvalue that would do a full reduction and not return indices, then yes that would apply.

**Comment by michaelklachko** [comment]
I see. What what the reason to fix this issue? Are there examples of where the old behavior could lead to training instability? What should I do if I have to backprop through kthvalue during training and I want to spread the gradients evenly?

**Comment by albanD** [comment]
Hi,

Instability is a strong word but unexpected behaviors. In particular because the value that was getting all the gradient was chosen in a non-deterministic way and could change across devices.

But also from a more principled point of view. When computing subgradients, we prefer the one with minimum norm (as it is always a descent direction). So in this case, the even distribution across all inputs that realize the value.

To do the same thing with kthvalue I guess you will need to do it yourself:
```python
import torch
from torch import autograd

# No dim can be specified an only full reduction is done
class MyKthvalue(autograd.Function):
    @staticmethod
    def forward(ctx, inp, k):
        res = inp.contiguous().view(-1).kthvalue(k).values
        ctx.save_for_backward(inp, res)
        return res

    @staticmethod
    def backward(ctx, gO):
        inp, res = ctx.saved_tensors
        mask = (inp == res)
        count = mask.sum()
        gO = gO / count
        return mask * gO, None


a = torch.randint(0, 10, (4, 4, 4), dtype=torch.float, requires_grad=True)
k = 3

MyKthvalue.apply(a, k).sum().backward()

print(a)
print(a.grad)
```

**Comment by chanshing** [comment]
Does this change `ReLU` and related layers that depend on the `max` function? If so, what are the implications?

**Comment by zasdfgbnm** [comment]
@chanshing ReLU are not changed
